File: 0021-merge-two-sorted-lists/0021-merge-two-sorted-lists.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} list1
 * @param {ListNode} list2
 * @return {ListNode}
 */
var mergeTwoLists = function(list1, list2) {
    // create a new node
    // set current to tempNode
    // while list1 and list2 are not null
    // compare the values and add smaller value node to current
    // move the corresponding pointer to the next node
    // return merged list starting from tempNode.next so the 0 is not incl

    let tempNode = new ListNode(null);
    let current = tempNode;

    while (list1 && list2) {
        console.log(tempNode)
        if (list1.val <= list2.val) {
            current.next = list1;
            list1 = list1.next
        } else {
            current.next = list2;
            list2 = list2.next;
        }
        current = current.next;
    }
    current.next = list1 || list2;

    return tempNode.next; 
    
};

File: 0021-merge-two-sorted-lists/0021-merge-two-sorted-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        dummyHead = ListNode(None)
        tail = dummyHead
        current1 = list1
        current2 = list2

        while current1 is not None and current2 is not None:
            if current1.val < current2.val:
                tail.next = current1
                current1 = current1.next
            else:
                tail.next = current2
                current2 = current2.next

            tail = tail.next
        
        if current1 is None:
            tail.next = current2
        
        if current2 is None:
            tail.next = current1
        
        return dummyHead.next

                